zeebe/docs/rolling_updates.md

```diff
@@ -1,5 +1,76 @@
+
+## Breaking Changes
+
+To support rolling updates, we have to ensure that all patch version of the previous or current minor are compatible with the development version.
+
+### Network
+
+Old and new versions need to be able to communicate with each other.
+Most importantly, they need to be able to form a raft replication group, and update cluster membership via SWIM.
+The command API should also remain compatible between old and new versions but exceptions can be made.
+Breaking command API compatibility drastically increases the impact of rolling updates because it may block all processing until the rolling update is finished.
+
+Compatibility can be achieved through differently, depending on the use case.
+For example:
+1. New version supports both old and new protocol and can automatically choose which to use.
+2. New version supports a new protocol but doesn't use it by default.
+3. New version detects an old version and has explicit update logic.
+
+The first option is the classic solution to evolve network protocols.
+To make this possible, we recommend to always version protocols, for example by prefixing all messages with a single integer version.
+This makes it possible for old versions to detect newer protocol versions and for new versions to support both protocol versions at the same time.
+See `AbstractMessageDecoder` and its implementations for an (old) example.
+Details are documented in our [messaging](./messaging.md#protocol-format) docs.
+
+The second option should be avoided if possible because it requires more effort.
+When switching to a new protocol requires a configuration change, this has a higher impact than the rolling update itself.
+Additionally, the old protocol has to be deprecated and can only be removed later.
+When removing a deprecated protocol, there is the risk that unaware users encounter problems during the rolling update.
+
+The third option is the one we chose when implementing dynamic scaling where new versions would communicate via a new gossip protocol that old versions did not support.
+The new version avoids the need to communicate with an old version because it can assume the state of the old version and make decisions without communicating.
+See `RollingUpdateAwareInitializerV83ToV84` for implementation details.
+
+### Processing
+
+Processing is ideally compatible between old and new versions.
+This is not always possible, for example because the new version supports a new feature or has changed the behavior of an existing feature.
+In that case we must fail early and stay on the safe side by stopping processing.
+
+There are two mechanisms that ensure processing correctness:
+1. Processing and replay is stopped when encountering commands or events that are unknown.
+2. Events have a version that is incremented whenever the behavior changes.
+
+The event versioning is not automatically enforced and requires awareness of the developers.
```

Comment:
💭  should we consider enforcing it?

Reply:
I don't think that's really possible because we can't automatically detect whether the behavior is different or not.

Reply:
@megglos Instead of enforcing it, we can design the code to avoid not noticing that a change actually changes behavior of previously written events, like:
- https://github.com/camunda/zeebe/issues/13813